### PR TITLE
Issue 43: IP validation added to the heartbeat

### PR DESCRIPTION
**Pull request by Peterburnett**
Settings menu contains entry for the IP validation

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -85,6 +85,7 @@ function check_climaintenance($configfile) {
 
 $status = "";
 
+
```
Comment (brendanheywood) [inline index.php]: just watch whitespace in diffs

File: index.php [hunk 2]
```diff
@@ -115,6 +116,13 @@ function failed($reason) {
 require($CFG->dirroot . '/lib/setup.php');
 require_once($CFG->libdir.'/filelib.php');
 
+//IP Locking, check for remote IP in validated list, if not, exit
+$allowedips = get_config('tool_heartbeat','ipconfig');
+if ((!(remoteip_in_list($allowedips))) && (trim($allowedips) !== '')){
+    print "IP is not in validated range";
```
Comment (brendanheywood) [inline index.php]: There is a core api for notications:

echo $OUTPUT->notification($errormsg);

File: index.php [hunk 3]
```diff
@@ -115,6 +116,13 @@ function failed($reason) {
 require($CFG->dirroot . '/lib/setup.php');
 require_once($CFG->libdir.'/filelib.php');
 
+//IP Locking, check for remote IP in validated list, if not, exit
```
Comment (brendanheywood) [inline index.php]: this is a bit of a gotcha with this plugin, there are muliple endpoints, ie index.php or croncheck.php etc I would refactor this check into one place and then reuse this across all of them

note that not all the end points need this check, just the ones which expose some sort of info that could be leveraged which I think is just index.php and croncheck.php and maybe upload.php

Comment (Peterburnett) [reply index.php]: Sure, missed the multiple endpoints

File: index.php (hunk 2 above)
Comment (brendanheywood) [reply index.php]: ah sorry you are already using this below in the gui, in this context I think what you did is good. Maybe the only improvement could be sending a http 403 instead

File: croncheck.php [hunk 4]
```diff
@@ -32,6 +32,8 @@
 
 define('NO_UPGRADE_CHECK', true);
 
+include('iplock.php');
```
Comment (brendanheywood) [inline croncheck.php]: this is fine as is, but it's probably a better habit to always use require_once() for things which are libs and it's also more 'moodley'

File: croncheck.php [hunk 5]
```diff
@@ -117,6 +119,13 @@
     header('Expires: Tue, 04 Sep 2012 05:32:29 GMT');
 }
 
+// IP Locking, check for remote IP in validated list, make sure not run from CLI, if not, exit    $allowedips = get_config('tool_heartbeat','ipconfig');
```
Comment (brendanheywood) [inline croncheck.php]: I'd refactor this whole chunk of code into iplock so this is just a one liner like:

check_ip();

File: iplock.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * Validates any remote IP connected to a page calling this script, and checks
+ * against current configs in plugin settings.
+ * 
+ * @param string list of safe IP addresses to validate the remote IP against.
+ * @return bool returns true if remote IP matches against the safe list, or the safe list is empty
+ * 
+ */
+
+function validate_IP_against_config($iplist) {
+    if (remoteip_in_list($iplist)){
```
Comment (brendanheywood) [inline iplock.php]: see above, I'd refactor this to do the check and throw the 403 when it fails or do nothing otherwise

If this was any more complex I'd split it into two functions, one which checks and returns true / false, and the second which calls it and then does the 403 - this is so the bulk of it could be unit tested. but this stuff is relatively trivial so I think that is overkill right now

File: iplock.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        return true;
+    } else if (trim($iplist) == ''){
+        return true;
+    } else {
+        return false;
+    }
+}
```
Comment (brendanheywood) [inline iplock.php]: watch the trailing lack of new lines in files, this should be an IDE setting somewhere. every line should have a \n at the end

File: lang/en/tool_heartbeat.php [hunk 8]
```diff
@@ -30,6 +30,13 @@
 $string['progresshelp'] = 'This tests that all the various output buffers in the entire stack are corrent including but not limited to php, ob, gzip/deflat, varnish, nginx etc';
 $string['testwarning'] = 'Fake a warning';
 $string['testerror'] = 'Fake a critical';
+$string['ipconfig'] = 'IP Config';
```
Comment (brendanheywood) [inline lang/en/tool_heartbeat.php]: this is not your fault at all, just an fyi but generally it's nice if the language string files are sorted by key.

If you find that all your new keys are spread out it's probably a sign that they have inconsistent names and maybe should have a common prefix to group them nicely.

File: index.php [hunk 9]
```diff
@@ -115,6 +118,13 @@ function failed($reason) {
 require($CFG->dirroot . '/lib/setup.php');
 require_once($CFG->libdir.'/filelib.php');
 
+// IP Locking, check for remote IP in validated list, if not, exit
+$allowedips = get_config('tool_heartbeat','ipconfig');
```
Comment (brendanheywood) [inline index.php]: be careful with naming things, eg here we have a setting called 'ipconfig' which isn't self explanatory, and it's not clear whether it's a whitelist or black list or something else.

You set it to a var called $allowedips so the inconsistency jumps out, I think they should have the same name in this particular case

File: iplock.php [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ * 
+ * @param string list of safe IP addresses to validate the remote IP against.
+ * @return bool returns true if remote IP matches against the safe list, or the safe list is empty
+ * 
+ */
+
+function validate_IP_against_config($iplist) {
+    if (remoteip_in_list($iplist)){
+        return true;
+    } else if (trim($iplist) == ''){
```
Comment (brendanheywood) [inline iplock.php]: this is a potential gotcha, eg get_config() might return false or null if there is no setting which can happen if you have a fresh install and the admin settings hasn't explicitly been set to an empty string yet. It's easy to have subtle edge cases depending on the comparator, I think in this case you are ok here but I'd test it

File: croncheck.php (hunk 5 above)
Comment (Peterburnett) [reply croncheck.php]: Currently unable to put the get_config inside of IP Lock, causes a bug with cron on the croncheck web execution. Tried it for a while, couldn't get it to play nice

File: croncheck.php [hunk 11]
```diff
@@ -31,6 +31,15 @@
...
 define('NO_UPGRADE_CHECK', true);
+// @codingStandardsIgnoreStart
+if (isset($argv)) {
+    define('CLI_SCRIPT', true);
+} else {
+    define('CLI_SCRIPT', false);
+}
+//@codingStandardsIgnoreEnd
+require('../../../config.php');
+require_once('iplock.php');
```
Comment (brendanheywood) [inline croncheck.php]: I thnk this code diff chunk isn't needed, we only need to iplock function if we actually check for ip's and we don't need to do that at all if we run this from the cli.

So I think the require_once and then the check itself can be moved to line 103 to just after where the config.php was originally
